# California Policy Lab Data Analyst Skills Assessment
by Victoria Hollingshead

In [ ]:
# For this exercise, I am using Python. Below are the libraries I used to complete this task.

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn

In [2]:
# Import data
link_testa_df = pd.read_csv('/Users/victoriahollingshead/Documents_Secure/CPL_2025/cpl_da_skills_assessment/Link_datasets/Link_testA.csv')
link_testb_df = pd.read_csv('/Users/victoriahollingshead/Documents_Secure/CPL_2025/cpl_da_skills_assessment/Link_datasets/Link_testB.csv')

## Exploratory Data Analysis

In [5]:
# Basic Information for Link Test A
link_testa_df.info()            
display(link_testa_df.head())      
display(link_testa_df.tail())      
display(link_testa_df.sample(5))   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2216 entries, 0 to 2215
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   first_name  2216 non-null   object 
 1   last_name   2216 non-null   object 
 2   street_num  2216 non-null   object 
 3   address_1   2216 non-null   object 
 4   address_2   2216 non-null   object 
 5   suburb      2216 non-null   object 
 6   postcode    2216 non-null   int64  
 7   state       2216 non-null   object 
 8   DOB         2216 non-null   object 
 9   SSN         2200 non-null   float64
dtypes: float64(1), int64(1), object(8)
memory usage: 173.2+ KB


first_name  last_name street_num         address_1             address_2  \
0   nicholas     clarke         13     gaylard place        tryphinia view   
1        mya     maksim         41   lambrigg street            brest hill   
2    katelyn   campbell         11       ogden close     willandra village   
3    makayla    herbert          5    plummer street           tarwyn park   
4        ben   armiento         17     burnell place   john flynn hospital   

            suburb  postcode state       DOB        SSN  
0   wetherill park      2810   nsw  19041223  3927795.0  
1         st kilda      6050   vic  19650630  1488918.0  
2          preston      2148   tas  19291114  4043690.0  
3         hopetoun      3925   vic  19111103  1756031.0  
4        fairlight      3630   qld            8511174.0

first_name      last_name street_num             address_1  \
2211     samuel          miles        138   hawkesbury crescent   
2212       jack      bridgland        170        thermeda place   
2213      lucas       portlock          2           mccaw place   
2214       tess      battisson          1           wyola place   
2215     amelia   graham-jones          5     mcgivern crescent   

                  address_2            suburb  postcode state       DOB  \
2211                               north ward      2148    sa             
2212              mirreanda     berkeley vale      2286   nsw  19980721   
2213              greenwood        ascot vale      2680   nsw  19960325   
2214                 malsah   west wollongong      2210   nsw  19230918   
2215   australian fair west          new town      2163   act  19000619   

            SSN  
2211  2763477.0  
2212  2930436.0  
2213  7237519.0  
2214  3152618.0  
2215  5013227.0

first_name  last_name street_num           address_1  \
627     cassidy   morrison         13    a'beckett street   
1437       alex      hanna        146       pocket avenue   
1573      paris      green         19         zamia place   
263      daniel      mcard         34   maccallum circuit   
685      emiily       mees         31     campbell street   

                address_2      suburb  postcode state       DOB        SSN  
627                argyle      ardeer      2518   nsw  19160403  6156712.0  
1437           scottsdale    yangebup      3941   vic  19040515  7193215.0  
1573            glendower   frankston      5159   qld  19291117  7101331.0  
263    westwood homestead    homebush      2619   nsw  19920927  2550366.0  
685               lingard      elwood      3805    wa  19301201  4611758.0

In [9]:
# Basic Information for Link Test B
link_testb_df.info()            
display(link_testb_df.head())      
display(link_testb_df.tail())      
display(link_testb_df.sample(5))  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   given_name     2200 non-null   object
 1   surname        2200 non-null   object
 2   street_number  2200 non-null   object
 3   address_1      2200 non-null   object
 4   address_2      2200 non-null   object
 5   suburb         2200 non-null   object
 6   postcode       2200 non-null   int64 
 7   state          2200 non-null   object
 8   date_of_birth  2200 non-null   object
 9   soc_sec_id     2200 non-null   int64 
dtypes: int64(2), object(8)
memory usage: 172.0+ KB


given_name    surname street_number           address_1  \
0      phoebe                       33          kumm place   
1       angip     wimmre                 baracchi crescent   
2   blakeston       lock            59        bonnor close   
3         ben       grub             9   girrahween street   
4        elle   mattehws            54       maddock place   

            address_2           suburb  postcode state date_of_birth  \
0    the manor garden      harris park      2323   vic      19061113   
1   mawarra herefords          marsden      2406    sa      19700928   
2                             ment one      7140    wa      19590930   
3                loma   brunswick west      1382   nsw      19360711   
4    glenburnie lodge    park orhcards      6058   vic      19960094   

   soc_sec_id  
0     3155906  
1     3795146  
2     6335558  
3     6198116  
4     9253952

given_name    surname street_number            address_1  \
2195      xavier   mogavero            10       epenarra close   
2196    morrison     sophie            11      waller crescent   
2197      bailey    stanley            30        colmer street   
2198     caitlin      white           174   collyburlc rescent   
2199   annabelle    conzens            27    mcluckie crescent   

                            address_2           suburb  postcode state  \
2195                           ramuda   west wolongong      3148   vic   
2196                    hmas cerberus         loganlea      3175   vic   
2197                          roclkea        rowew bay       850   nsw   
2198                                        kurwongbah      2456   nsw   
2199   earle haven retirement village                        870   vic   

     date_of_birth  soc_sec_id  
2195      19121210     2534734  
2196      19420302     7721195  
2197      19010804     3200988  
2198      19431207     2627787  
2199      19250508     9397671

given_name           surname street_number            address_1  \
1754    emii ly            cobgin             3          knox stteet   
2051     robert          westhorp            26     kosciusko avenue   
1250      heron            brooke           262   verbrugghens treet   
1202     cooper              webb            34      namadgi circuit   
368     lachlan   balasubramaniam            26    mackinolty street   

             address_2        suburb  postcode state date_of_birth  soc_sec_id  
1754                      camperdown      2230    wa      19991209     9931751  
2051         oa kp ark                    3109   vic      19690516     1463700  
1250       oxley's way       hackham      2560   qld                   7028101  
1202           fernale     newcastle      2566   nsw      19621205     8131556  
368    balmora lgarden   scarnorough      2388   nsw      19731007     1798069

### Observations

At first glance, it appears there are only 16 null values for SSN Link Test A, but after inspecting the head and tail of each dataset, I can see there are blank lines for various variables. This is also true for Link Test B. In addition, there are 16 fewer values in the Link Test B dataset than the List Test A dataset and a mismatch between variable names. This will require a more thorough review to understand which data cleaning steps may be necessary. 

In [ ]:
# Normally, I would divide columns into categorical and numerical values, but because postcode, DOB, and SSN are unique values, I will assess them along with the categorical variables. 
link_testa_df.columns

lista_variables = ['first_name', 'last_name', 'street_num', 'address_1', 'address_2', 'suburb', 'state', 'postcode', 'DOB', 'SSN']

In [17]:
for col in lista_variables:
    print(f"Column: \n {col}")
    print()
    print(f"Unique values: \n {link_testa_df[col].nunique()}")
    print()
    print(f"Value counts: \n {link_testa_df[col].value_counts()}")
    print()
    print(f"Percentage of each unique value: \n{link_testa_df[col].value_counts(normalize=True)}")
    print()
    print(f"Sum of null values: \n{link_testa_df[col].isnull().sum()}" )
    
# For full outline of List A Summary statistics, refer to list_a_summary_stats.txt

Column: 
 first_name

Unique values: 
 596

Value counts: 
              54
 emiily      43
 joshua      35
 jack        32
 lachlan     31
             ..
 zakariah     1
 peta         1
 savannah     1
 iain         1
 billy        1
Name: first_name, Length: 596, dtype: int64

Percentage of each unique value: 
             0.024368
 emiily      0.019404
 joshua      0.015794
 jack        0.014440
 lachlan     0.013989
               ...   
 zakariah    0.000451
 peta        0.000451
 savannah    0.000451
 iain        0.000451
 billy       0.000451
Name: first_name, Length: 596, dtype: float64

Sum of null values: 
0
Column: 
 last_name

Unique values: 
 1057

Value counts: 
  white           61
 clarke          55
 green           42
 webb            40
 reid            37
                 ..
 kincaid          1
 gigney           1
 moolenaar        1
 zoanetti         1
 graham-jones     1
Name: last_name, Length: 1057, dtype: int64

Percentage of each unique value: 
 white        

### List A Summary Statistics

first_name
* Unique values: 596
* Nulls: 0
* Blanks: 54
* Data Cleaning Next Steps: Because there are only 596 unique values of ~2200 entires, first_name cannot be used alone as an identifier to join with List Test B on given_name. There are also several cases where traditional names are slightly misspelled (e.g., emiily for Emily, jeremf for Jeremy, harrion for Harrison). There may be an opportunity to use fuzzy logic to match the first or given name between List A and List B. 

last_name
* Unique values: 1,057
* Nulls: 0
* Blanks: 0
* Data Cleaning Next Steps: There are several repeated values (e.g., white (61), clarke (55), green (42)). 

street_num
* Unique values: 311
* Nulls: 0 
* Blanks: 72 (this makes up about ~3% of the dataset!)

address_1
* Unique values: 1,486
* Nulls: 0
* Blanks: 50 (this makes up ~2% of the dataset)

address_2
* Unique values: 1,476
* Nulls: 0
* Blanks: 190 (this makes up about 9% of the dataset!)

suburb
* Unique values: 1,108
* Nulls: 0
* Blanks: 20

state
* Unique values: 9
* Most common: "nsw" (746), followed by "vic" (537) and “qld” (413)
* Nulls: 0
* Blanks: 23
* Data Cleaning Next Steps: These states correspond to states and regions in Australia. After looking at each of the unique values, I can see that this column is fairly clean, as there are 8 states/territories and 1 blank value. I can fill out the blank states with a corresponding zip code if that data is present for the respective user. 

postcode
* Unique values: 1,029
* Most common: "4740" (15 times)
* Nulls: 0
* Blanks: 0

DOB
* Unique values: 2,106
* Nulls: 0
* Blanks: 41
* Data Cleaning Next Steps: 95% of DOBs are unique. This is great news, as it may be a key identifier in joining List A with List B. 

SSN
* Unique values: 2,200
* Nulls: 16
* Blanks: 0
* Data Cleaning Next Steps: This variable is also highly unique and can be a great key identifier between lists. 

In [15]:
link_testb_df.columns

listb_variables = ['given_name', 'surname', 'street_number', 'address_1', 'address_2', 'suburb', 'postcode', 'state', 'date_of_birth', 'soc_sec_id']

In [ ]:
for col in link_testb_df:
    print(f"Column: \n {col}")
    print()
    print(f"Unique values: \n {link_testb_df[col].nunique()}")
    print()
    print(f"Value counts: \n {link_testb_df[col].value_counts()}")
    print()
    print(f"Percentage of each unique value: \n{link_testb_df[col].value_counts(normalize=True)}")
    print()
    print(f"Sum of null values: \n{link_testb_df[col].isnull().sum()}" )
    
# For full outline of List B Summary statistics, refer to list_b_summary_stats.txt

### List B Summary Statistics

given_name
* Unique values: 1000
* Nulls: 0
* Blank: 91

surname
* Unique values: 1379
* Nulls: 0
* Blank: 52
* Data Cleaning Next Steps: There are move unique surnames in this dataset, suggesting there may be some inconsistency in spelling or formatting. 

street_number
* Unique values: 307
* Nulls: 0
* Blank: 121
* Data Cleaning Next Steps: Blanks make up 5.5% of the dataset. This is likely inconsequential to our analysis.

address_1
* Unique values: 1795
* Nulls: 0
* Blank: 92

address_2
* Unique values: 1589
* Nulls: 0
* Blank: 387 (this makes up about 17.6% of the dataset!)
* Data Cleaning Next Steps: It’s unlikely this variable will be much help in linking due to the high blank count. 

suburb
* Unique values: 1378
* Nulls count: 0
* Blank count: 45

postcode
* Unique values: 1138
* Nulls count: 0
* Blank count: 0

state
* Unique values: 33
* Nulls count: 0
* Blank count: 52
* Data Cleaning Next Steps: There are only 6 states and 2 territories in Australia, therefore, a level of standardization is necessary to ensure uniform spelling and formatting across entries. The state composition is similar to that in List A. 

date_of_birth
* Unique values: 2054
* Nulls count: 0
* Blank count: 79
* Data Cleaning Next Steps: Like in List A, DOB is relatively unique. 

soc_sec_id
* Unique values: 2200
* Nulls count: 0
* Blank count: 0
* Data Cleaning Next Steps: Like in List A, SSN is unique and clean.

### Data Cleaning Strategy

1. Standardize column names between List A and List B
2. Standardize the state column in List B by updating mispelled state abbreviations
3. Fill blank state values with corresponding zip code 

### Linkage Strategy
4. As a standard step, I will check how many perfect matches exist
5. Because SSN is highly unique and complete in both the List A and List B datasets, I will then join the lists on SSN
6. Assuming SSN columns are clean in both datasets, I'll then evaluate the merge based in conjunction with DOB, state and/or zip code values 

## Data Cleaning

In [47]:
# Rename all columns in List B to match with List A
link_testb_clean = link_testb_df.copy()

columns_rename = {
    'given_name': 'first_name',
    'surname': 'last_name',
    'street_number': 'street_num',
    'date_of_birth': 'DOB',
    'soc_sec_id': 'SSN'
}

link_testb_clean.rename(columns=columns_rename, inplace=True)
link_testb_clean.columns # check my work

Index(['first_name', 'last_name', 'street_num', 'address_1', 'address_2',
       'suburb', 'postcode', 'state', 'DOB', 'SSN'],
      dtype='object')

In [48]:
nsw_val = ["nsw", "nws", "nss", "nsq", "nse", "nww", "ndw"]
vic_val = ["vic", "vci", "vix", "vid", "vco", "vfc", "fic"]
qld_val = ["qld", "qdl", "qod", "qle"]
wa_val  = ["wa", "wq", "aw"]
sa_val  = ["sa", "sk", "sz", "sj"]
tas_val = ["tas", "tsa", "as"]
act_val = ["act"]
nt_val  = ["nt", "ni", "st"]

def state_mapping(state):
    state = str(state).strip().lower()
    if state in nsw_val:
        return nsw_val[0]
    elif state in vic_val:
        return vic_val[0]
    elif state in qld_val:
        return qld_val[0]
    elif state in wa_val:
        return wa_val[0]
    elif state in sa_val:
        return sa_val[0]
    elif state in tas_val:
        return tas_val[0]
    elif state in act_val:
        return act_val[0]
    elif state in nt_val:
        return nt_val[0]
    else:
        return "blank"

In [49]:
link_testb_clean["state"] = link_testb_clean["state"].apply(state_mapping)

In [50]:
link_testb_clean["state"].value_counts() # check my work

nsw      715
vic      545
qld      405
wa       221
sa       163
tas       58
blank     52
act       26
nt        15
Name: state, dtype: int64

In [71]:
# Next, I'll fill in blank states with corresponding value depending on zipcode. 
# I will use the Link_datasets/australian-postcodes-2021-04-23.csv file for this step. 
# All credit goes to authors from "https://github.com/EricHuynh10/australia-postcodes/blob/main/aus_postcode.csv" which compiled postcode data from the Australia Bureau of Statistics.

zipcode_aus = pd.read_csv('/Users/victoriahollingshead/Documents_Secure/CPL_2025/cpl_da_skills_assessment/Link_datasets/aus_postcode.csv')

zip_col_rename = {
    'state' : 'state_full',
    'state_code' : 'state',
}
zipcode_aus = zipcode_aus.rename(columns = zip_col_rename)
zipcode_aus["suburb"] = zipcode_aus["suburb"].str.lower()
zipcode_aus["state"] = zipcode_aus["state"].str.lower()
zipcode_aus.head()

suburb  postcode          state_full state
0  darwin city       800  Northern Territory    nt
1      jingili       810  Northern Territory    nt
2      leanyer       812  Northern Territory    nt
3  stuart park       820  Northern Territory    nt
4       rakula       822  Northern Territory    nt

In [72]:
link_testb_clean_add = pd.merge(link_testb_clean[["state", "postcode"]], zipcode_aus[["state", "postcode"]], left_on = "postcode", right_on = "postcode")

In [73]:
df_no_duplicates = link_testb_clean_add.drop_duplicates(subset=['postcode'], keep='first')
print(df_no_duplicates.head)

<bound method NDFrame.head of      state_x  postcode state_y
0        vic      2323     nsw
4         sa      2406     qld
6         wa      7140     tas
8        vic      6058      wa
9        qld      2026     nsw
...      ...       ...     ...
2091     qld      5163      sa
2092     nsw      5330      sa
2093      sa      4228     qld
2094   blank      5033      sa
2095     nsw       850      nt

[1043 rows x 3 columns]>


### Data Cleaning Observations

While I was able to clean the state column for List B, I could not further substantiate the dataset with zipcode + state combinations from the Australia Bureau of Statistics for additional cleaning.

## Linkage Analysis

In [ ]:
linkage_all = pd.merge(link_testb_clean, link_testa_df, how = "inner")
print(linkage_all.shape)
linkage_all.head(10)
linkage_all.tail(10)

# As expected, there are no two records that are a perfect match between List A and List B. 

(0, 10)


Empty DataFrame
Columns: [first_name, last_name, street_num, address_1, address_2, suburb, postcode, state, DOB, SSN]
Index: []

In [74]:
a_SSN = set(link_testa_df['SSN'])
b_SSN = set(link_testb_clean['SSN'])

In [75]:
common = a_SSN & b_SSN     # values in common
only_in_a = a_SSN - b_SSN   # values only in a_SSN
only_in_b = b_SSN - a_SSN   # values only in b_SSN
all_unique = a_SSN | b_SSN  # total unique values

# Summary
print(f"Total in a_SSN: {len(a_SSN)}")
print(f"Total in b_SSN: {len(b_SSN)}")
print(f"In common: {len(common)}")
print(f"Only in a_SSN: {len(only_in_a)}")
print(f"Only in b_SSN: {len(only_in_b)}")
print(f"Total unique values: {len(all_unique)}")

Total in a_SSN: 2216
Total in b_SSN: 2200
In common: 1005
Only in a_SSN: 1211
Only in b_SSN: 1195
Total unique values: 3411


### Linkage Observations based on SSN as key identifier

Only 50% of SSN values exist in both List A and List B. This runs contrary to my initial assumption that SSN would be a key identifier due to its uniqueness in both datasets. I will next evaluate the quality of these matches based on SSN alone. 

In [121]:
linkage_SSN = pd.merge(link_testb_clean, link_testa_df, how = "inner", on = "SSN")


In [122]:
print(linkage_SSN.shape)
linkage_SSN.head(10)
linkage_SSN.tail(10)
linkage_SSN.sample(10)

(1005, 19)


first_name_x last_name_x street_num_x                address_1_x  \
225           dam        hage            1   woodlands surgery centre   
774          kade        hand           14             morduant place   
849        jordan    campbell            3                              
181        oliver       berry           18             mcgill s treet   
282        tiarna     cannell          685               deacon close   
1004    annabelle     conzens           27          mcluckie crescent   
788         grace    sherriff           25          chuculba crescent   
167        damien        webb           34          molesworth street   
631        tomney       chloe           24             woollum street   
30         joshua     arribas           30            nicholas street   

                          address_2_x         suburb_x  postcode_x state_x  \
225                       sweet place   port macqyarie        3206     qld   
774                          collaroy         ringwood        3356     vic   
849                       lamont park          garbutt        2207     vic   
181                                            bayview        3082     nsw   
282                                      fishingupoint        2287      wa   
1004   earle haven retirement village                          870     vic   
788                           wunpaml        riverwood        2101     qld   
167              fairfield housing bg    werribeesouth        2484     qld   
631                          carranya           pymble        2321     nsw   
30                      dangaroo park           kirwan        4086      nt   

         DOB_x      SSN first_name_y last_name_y street_num_y  \
225   19430116  1799750          sam        hage            1   
774             5013199         kade       white           14   
849   19490816  3930891       jordan    campbell            3   
181   19720119  2589280       oliver       berry           18   
282   19020305  6547713       tiarna     cannell          685   
1004  19250508  9397671    annabelle     couzens           27   
788   19111215  2181571        paige    sherriff           25   
167   19440826  5682631       damien        webb           34   
631   19901223  8831246        chloe    lazaraki           24   
30    19800502  7705455       joshua     arribas           30   

             address_1_y                      address_2_y         suburb_y  \
225          sweet place         woodlands surgery centre   port macquarie   
774       morduant place                         collaroy         ringwood   
849     brownless street                      lamont park          garbutt   
181       arthaldo court                           kalang          bayview   
282         deacon close                                             wooli   
1004   mcluckie crescent   earle haven retirement village                    
788    chuculba crescent                          wunpalm      marks point   
167    molesworth street             fairfield housing bg   werribee south   
631       woollum street                                            pymble   
30       nicholas street                    dangaroo park         narellan   

      postcode_y state_y     DOB_y  
225         3206     qld  19430116  
774         3356     vic            
849         2207     vic  19490816  
181         3802     nsw  19720119  
282         2287      wa  19020305  
1004         870     vic  19260508  
788         2101     qld  19111215  
167         2448     qld  19440826  
631         2321     nsw  19901223  
30          4068      nt  19800502

At first glance, applying SSN as the key identifier decently matches List A and List B. I notice some discrepencies, like the reversal of first_name and last_name, general mispellings, and post code discrepencies, I will attempt to quantify the match by checking for commonality in name and DOB. 

In [141]:
from collections import Counter # count the occurences of elements in an iterable

# Combine first and last name to compare by character
linkage_SSN['full_name_x'] = linkage_SSN['first_name_x'].str.strip() + linkage_SSN['last_name_x'].str.strip()
linkage_SSN['full_name_y'] = linkage_SSN['first_name_y'].str.strip() + linkage_SSN['last_name_y'].str.strip()

# Clean DOB columns
linkage_SSN['DOB_x'] = linkage_SSN['DOB_x'].astype(str).str.strip() 
linkage_SSN['DOB_y'] = linkage_SSN['DOB_y'].astype(str).str.strip()

# Calculate character percent match in full_name
def char_percent_match(str1, str2):
    str1 = str(str1).lower()
    str2 = str(str2).lower()
    c1 = Counter(str1)
    c2 = Counter(str2)
    common = c1 & c2
    common_total = sum(common.values())
    total_length = len(str1) + len(str2)
    return round((2 * common_total / total_length) * 100) if total_length > 0 else 0 # This provides a normalized score

# Apply to names and DOB
linkage_SSN['common_char_percent'] = linkage_SSN.apply(
    lambda row: char_percent_match(row['full_name_x'], row['full_name_y']),
    axis=1
)

linkage_SSN['dob_match'] = linkage_SSN['DOB_x'] == linkage_SSN['DOB_y'] # This provides a True of False

linkage_SSN['dob_char_percent'] = linkage_SSN.apply(
    lambda row: char_percent_match(row['DOB_x'], row['DOB_y']),
    axis=1
)

linkage_SSN.head()


first_name_x  last_name_x street_num_x         address_1_x   address_2_x  \
0     victoria   zimmermann           37   o'sullivan street      kelgoola   
1         lara       stirpe           60     wilshire street                 
2         liam       pascoe          311      tillyadr drive   lindenleigh   
3        teaha     kyriacou           17          iron place                 
4         sedn         lgke           30    glossop crescent     craigston   

      suburb_x  postcode_x state_x     DOB_x      SSN  ...      address_2_y  \
0         nyah        2842      sa  19240517  7803602  ...         kelgoola   
1     wanneroo        2106      wa            7108358  ...   the boulevarde   
2     hamilton        4030     nsw  19960928  2933911  ...      lindenleigh   
3   tambo pper        2783     nsw  19720829  6810695  ...                    
4        megan        4812     vic  19980906  7108946  ...        craigston   

       suburb_y postcode_y state_y     DOB_y         full_name_x  \
0          nyah       2842     qld  19240517  victoriazimmermann   
1      wanneroo       2160      wa  19100118          larastirpe   
2      hamilton       4030     nsw  19960928          liampascoe   
3   tambo upper       2783     nsw  19720829       teahakyriacou   
4         megan       4812     vic  19980906            sednlgke   

          full_name_y common_char_percent dob_match dob_char_percent  
0  victoriazimmermann                 100      True              100  
1          larastirpe                 100     False                0  
2        baileypascoe                  82      True              100  
3       teahakyriacou                 100      True              100  
4            lukeseng                  88      True              100  

[5 rows x 24 columns]

At this point, I'm ready to divide this dataset into 4 groups based on a threshold of 80% commonality: 

- High common_char_percentage (>80%), High dob_char_percent (>80%)
- High common_char_percentage (>80%), Low dob_char_percent (<80%)
- Low common_char_percentage (<80%), High dob_char_percent (>80%)
- Low common_char_percentage (<80%), Low dob_char_percent (<80%)

In [142]:
# common_char_percentage vs dob_char_percentage

high_high = (linkage_SSN['common_char_percent'] > 80) & (linkage_SSN['dob_char_percent'] > 80)
high_low = (linkage_SSN['common_char_percent'] > 80) & (linkage_SSN['dob_char_percent'] < 80)
low_high = (linkage_SSN['common_char_percent'] < 80) & (linkage_SSN['dob_char_percent'] > 80)
low_low = (linkage_SSN['common_char_percent'] < 80) & (linkage_SSN['dob_char_percent'] < 80)

eval_list = [high_high, high_low, low_high, low_low]

linkage_SSN[high_high].head() # check my work

first_name_x  last_name_x street_num_x          address_1_x   address_2_x  \
0     victoria   zimmermann           37    o'sullivan street      kelgoola   
2         liam       pascoe          311       tillyadr drive   lindenleigh   
3        teaha     kyriacou           17           iron place                 
4         sedn         lgke           30     glossop crescent     craigston   
5       sophie        furze          590   goldsborough close                 

          suburb_x  postcode_x state_x     DOB_x      SSN  ...  \
0             nyah        2842      sa  19240517  7803602  ...   
2         hamilton        4030     nsw  19960928  2933911  ...   
3       tambo pper        2783     nsw  19720829  6810695  ...   
4            megan        4812     vic  19980906  7108946  ...   
5   carina heights        3171     qld  19110502  4765244  ...   

           address_2_y         suburb_y postcode_y state_y     DOB_y  \
0             kelgoola             nyah       2842     qld  19240517   
2          lindenleigh         hamilton       4030     nsw  19960928   
3                           tambo upper       2783     nsw  19720829   
4            craigston            megan       4812     vic  19980906   
5   tallageris-demijoh   carina heights       3171     qld  19110502   

          full_name_x         full_name_y common_char_percent dob_match  \
0  victoriazimmermann  victoriazimmermann                 100      True   
2          liampascoe        baileypascoe                  82      True   
3       teahakyriacou       teahakyriacou                 100      True   
4            sednlgke            lukeseng                  88      True   
5         sophiefurze         sophiefurze                 100      True   

  dob_char_percent  
0              100  
2              100  
3              100  
4              100  
5              100  

[5 rows x 24 columns]

In [143]:
total = linkage_SSN.shape[0]

# Total count 
print(linkage_SSN[high_high].shape[0])
print(linkage_SSN[high_low].shape[0])
print(linkage_SSN[low_high].shape[0])
print(linkage_SSN[low_low].shape[0])
print()


# Normalized count

print(linkage_SSN[high_high].shape[0]/total)
print(linkage_SSN[high_low].shape[0]/total)
print(linkage_SSN[low_high].shape[0]/total)
print(linkage_SSN[low_low].shape[0]/total)


777
69
144
11

0.7731343283582089
0.06865671641791045
0.14328358208955225
0.010945273631840797


In [144]:
# - High common_char_percent (>80%), High dob_char_percent (>80%)

SSN_high_high = linkage_SSN[high_high].copy()

# In this group, how many are perfect matches?
perfect_match = (SSN_high_high["common_char_percent"] == 100) & (SSN_high_high["dob_match"] == True)
SSN_high_high[perfect_match].shape[0]

# Percentage of total
SSN_high_high[perfect_match].shape[0]/linkage_SSN[high_high].shape[0]

0.676962676962677

In [ ]:
# What issues are causing the non-perfect match for this group?

# Cases where character count was < 100
less_than_100 = SSN_high_high[SSN_high_high["common_char_percent"] < 100][['full_name_x', 'full_name_y', 'DOB_x', 'DOB_y', 'common_char_percent', 'dob_match', 'dob_char_percent']]
SSN_high_high[SSN_high_high["common_char_percent"] < 100][['full_name_x', 'full_name_y', 'DOB_x', 'DOB_y', 'common_char_percent', 'dob_match', 'dob_char_percent']].sample(10)


full_name_x     full_name_y     DOB_x     DOB_y  common_char_percent  \
933       shaileung        shaileon  19230106  19230106                   82   
260   scriplpsmegan    meganscripps  19970108  19970108                   96   
564       tobproles       tobyroles  19770325  19770325                   89   
306     karlawilins    karlawilkins  19590520  19590520                   96   
702     emiilbishop    emiilybishop  19730322  19730322                   96   
792   hanahmorrison  hannahmorrison  19010126  19010126                   96   
634  harrietclalrke   harrietclarke  19590810  19590810                   96   
547     chlseablake    chelseablake  19770311  19770311                   96   
255    john-paulpra   john-paulparr  19800811  19800811                   96   
449       joewhitle        joewhite  19851122  19851122                   94   

     dob_match  dob_char_percent  
933       True               100  
260       True               100  
564       True               100  
306       True               100  
702       True               100  
792       True               100  
634       True               100  
547       True               100  
255       True               100  
449       True               100

In [153]:
# Cases where dob match was False
SSN_high_high[SSN_high_high["dob_match"] == False][['full_name_x', 'full_name_y', 'DOB_x', 'DOB_y', 'common_char_percent', 'dob_match', 'dob_char_percent']].sample(10)


full_name_x         full_name_y     DOB_x     DOB_y  \
828           lexiemyles          lexiemyles  19600069  19600609   
532      mccarthymikaela     mikaelamccarthy  19821001  19811001   
371     briannablackwell    briannablackwell  19430515  19440515   
450           jskstanbur        jackstanbury  19381131  19381113   
217         joshuaclarke        joshuaclarke  19011011  19011101   
367         jessfedonkin        jessiedonkin  19900096  19900906   
529          josephineyu         josephineyu  19310920  19110903   
73           liamrollins         liamrollins  19690525  19690515   
1004    annabelleconzens    annabellecouzens  19250508  19260508   
202   brooklynmercoerlla  brooklynmercorella  19050715  19050705   

      common_char_percent  dob_match  dob_char_percent  
828                   100      False               100  
532                   100      False                88  
371                   100      False                88  
450                    82      False               100  
217                   100      False               100  
367                    92      False               100  
529                   100      False                88  
73                    100      False                88  
1004                   94      False                88  
202                   100      False                88

In [ ]:
#- High common_char_percentage (>80%), Low dob_char_percent (<80%)

SSN_high_low = linkage_SSN[high_low].copy()
SSN_high_low[['full_name_x', 'full_name_y', 'DOB_x', 'DOB_y', 'common_char_percent', 'dob_match', 'dob_char_percent', 'postcode_x', 'postcode_y']]

# Perhaps adding one more match for postcode and suburb will help parse out matches for this group. Due to the Sr/Jr naming convention within families, you could have records with the same name but different DOBs. 
# As a next step, I may parse the DOB by year, month, and day to separate typo errors from users who may be generations apart. 

full_name_x        full_name_y     DOB_x     DOB_y  \
1           larastirpe         larastirpe            19100118   
6        phoebewhisson      phoebewhisson            19870307   
16   flytnnfitzpagrick   flynnfitzpatrick  19540125  19550713   
18     pearsonshepherd    pearsonshepherd  19020705  19320517   
33               james              james                       
..                 ...                ...       ...       ...   
923        elenibartel        elenibartel  19090725  19491112   
952  jamiehashemizadeh  jamiehashemizadeh  19851030  19231030   
979        laurenwhite        laurenwhite  19020813             
980      matthe wserin       erinmatthews  19351004  19960529   
984      gregorybujtas      gregorybujtas  19811017  19581002   

     common_char_percent  dob_match  dob_char_percent  postcode_x  postcode_y  
1                    100      False                 0        2106        2160  
6                    100      False                 0        2770        2770  
16                    97      False                75        3033        3033  
18                   100      False                75        2650        2650  
33                   100       True                 0        5203        5203  
..                   ...        ...               ...         ...         ...  
923                  100      False                50        2063        2063  
952                  100      False                75        4205        4205  
979                  100      False                 0        3041        3041  
980                   96      False                50        3160        3160  
984                  100      False                62        4285        4285  

[69 rows x 9 columns]

In [ ]:
# - Low common_char_percentage (<80%), High dob_char_percent (>80%)

SSN_low_high = linkage_SSN[low_high].copy()
SSN_low_high[['full_name_x', 'full_name_y', 'DOB_x', 'DOB_y', 'common_char_percent', 'dob_match', 'dob_char_percent', 'postcode_x', 'postcode_y']]

# These are likely cases where someone changed their first or last name (due to marriage, etc.), or one of the names were omitted from either List A or List B. 
# The matching algorithm could be updated such that the common_char_percent accounts for the case where only 1 name is present to avoid depressing the percentage.

full_name_x     full_name_y     DOB_x     DOB_y  common_char_percent  \
11          kyleryna      alissaryan  19420405  19420405                   56   
20       jessicalowe  jessicamillane  19120708  19120708                   72   
22       sylviewhite    sylviemorcom  19580901  19580901                   52   
26        abbeyrobsn   matthewrobson  19441016  19441016                   61   
29   katharinestubbs     izaacstubbs  19511101  19511101                   69   
..               ...             ...       ...       ...                  ...   
971         maimoody       mianguyen  19350114  19350114                   47   
982         liamgani    charlizegani  19831203  19830412                   70   
985      devanottens    joshuaottens  19410514  19410514                   61   
993          swoboda  lachlanswoboda  19931215  19931215                   67   
999             webb  jacksonbarisic  19450822  19450822                   11   

     dob_match  dob_char_percent  postcode_x  postcode_y  
11        True               100        2884        2880  
20        True               100        7279        7275  
22        True               100        4551        4551  
26        True               100        4000        4000  
29        True               100        5290        5290  
..         ...               ...         ...         ...  
971       True               100        5608        5608  
982      False                88        5013        5013  
985       True               100        5163        5163  
993       True               100        4740        4740  
999       True               100        2286        2286  

[144 rows x 9 columns]

In [ ]:
# - Low common_char_percentage (<80%), Low dob_char_percent (<80%)

SSN_low_low = linkage_SSN[low_low].copy()
SSN_low_low[['full_name_x', 'full_name_y', 'DOB_x', 'DOB_y', 'common_char_percent', 'dob_match', 'dob_char_percent', 'postcode_x', 'postcode_y']]


full_name_x       full_name_y     DOB_x     DOB_y  \
62       taliafeast        hollyfeast                       
115    ryleheverett       rylehsleath            19550126   
140     heronbrooke       aliciaheron            19320131   
272          clarkd     caitlynclarke  19700607  19420124   
364   alysahahyland      davidcandida  19731211  19781104   
511     harveywhite     harveyroselli                       
675  kynansincovich  matthewsincovich  19470922  19600731   
774        kadehand         kadewhite                       
781      lodgeadele      gillianlodge  19041015  19000626   
950       lkuequast         jadequast  19760328  19390121   
957   sybellameaney     madisonmeaney            19731218   

     common_char_percent  dob_match  dob_char_percent  postcode_x  postcode_y  
62                    60       True                 0        6063        6063  
115                   61      False                 0        2484        2484  
140                   45      False                 0        2560        2560  
272                   53      False                38        2090        2090  
364                   40      False                62        3146        3146  
511                   67       True                 0        3089        3089  
675                   67      False                50        4222        4221  
774                   59       True                 0        3356        3356  
781                   64      False                50        2765        2765  
950                   67      False                62        6210        6210  
957                   62      False                 0        3690        3690

### Conclusion: Linkage based on SSN as key identifier

- High common_char_percentage (>80%), High dob_char_percent (>80%)
  - Nearly 68% of the records in this category are a perfect match. 
- High common_char_percentage (>80%), Low dob_char_percent (<80%)
- Low common_char_percentage (<80%), High dob_char_percent (>80%)
- Low common_char_percentage (<80%), Low dob_char_percent (<80%)

### Linkage Observations based on non-SSN as key identifier

In [ ]:
linkage_nonSSN_b = link_testb_clean.copy()
linkage_nonSSN_a = link_testa_df.copy()

# if first and last name, postcode, and DOB is the same, and SSN is > 80% similar, these may also be viable matches
linkage_nonSSN_b['match_column'] = linkage_nonSSN_b['first_name'].str.strip() + linkage_nonSSN_b['last_name'].str.strip() + linkage_nonSSN_b['postcode'].astype(str).str.strip()+ linkage_nonSSN_b['DOB'].astype(str).str.strip()
linkage_nonSSN_a['match_column'] = linkage_nonSSN_a['first_name'].str.strip() + linkage_nonSSN_a['last_name'].str.strip() + linkage_nonSSN_a['postcode'].astype(str).str.strip()+ linkage_nonSSN_a['DOB'].astype(str).str.strip()

linkage_nonSSN_outer = pd.merge(linkage_nonSSN_b, linkage_nonSSN_a, how = "inner", on = "match_column")

In [175]:
linkage_nonSSN_outer.head()

first_name_x  last_name_x street_num_x          address_1_x   address_2_x  \
0     victoria   zimmermann           37    o'sullivan street      kelgoola   
1        kylie      seppelt            4        tannerip lace     wearmouth   
2        teaha     kyriacou           17           iron place                 
3     madeline        jolly            6                        crown allot   
4       sophie        furze          590   goldsborough close                 

          suburb_x  postcode_x state_x     DOB_x    SSN_x  ... first_name_y  \
0             nyah        2842      sa  19240517  7803602  ...     victoria   
1          modbury        4883     qld  19251024  3972479  ...        kylie   
2       tambo pper        2783     nsw  19720829  6810695  ...        teaha   
3      lilli pilli        2546     nsw  19280509  8473152  ...     madeline   
4   carina heights        3171     qld  19110502  4765244  ...       sophie   

   last_name_y street_num_y          address_1_y          address_2_y  \
0   zimmermann           37    o'sullivan street             kelgoola   
1      seppelt            4         tanner place            wearmouth   
2     kyriacou           17           iron place                        
3        jolly            6                               crown allot   
4        furze          590   goldsborough close   tallageris-demijoh   

          suburb_y postcode_y  state_y     DOB_y      SSN_y  
0             nyah       2842      qld  19240517  7803602.0  
1          modbury       4883      qld  19251024  3972497.0  
2      tambo upper       2783      nsw  19720829  6810695.0  
3      lilli pilli       2546      nsw  19280509  9133348.0  
4   carina heights       3171      qld  19110502  4765244.0  

[5 rows x 21 columns]

In [ ]:
linkage_nonSSN_outer.shape[0] # there are 434 cases where first and last name, postcode, and DOB are the same

# Remove cases where SSN_x and SSN_y are the same
linkage_nonSSN_outer[linkage_nonSSN_outer["SSN_x"] == linkage_nonSSN_outer["SSN_y"]].shape[0] # there are 384 cases where SSN is the same (covered in our previous analysis). To verify if the remaining ~50 cases are matches will take an additional review of SSN similarity. 

384